lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp

```diff
@@ -358,6 +358,133 @@ struct ConvolutionToConv2dPattern
                              adaptor.getOperandConstraints());
 
     rewriter.replaceOp(op, output);
+    return success();
+  }
+};
+//===----------------------------------------------------------------------===//
+// Gather Pattern Matching
+//===----------------------------------------------------------------------===//
+
+struct GatherToEmbeddingConversionPattern
+    : public OpConversionPattern<ttir::GatherOp> {
+  using OpConversionPattern<ttir::GatherOp>::OpConversionPattern;
+
+  LogicalResult checkBasicLegality(ttir::GatherOp op,
```

Comment:
I feel we need a better explanation on legality when are we able to lower from GatherOp to EmbeddingOp. Our EmbeddingOp is quite different semantically compared to GatherOp. I understand that we add bunch of constraints below which will ensure legality. But it's hard to understand where they exactly come from. I propose that you write a top level description for `checkBasicLegality` which will explain when we can use EmbeddingOp.

Reply:
I added a description above checkBasicLegality to explain the constraints.